kolibri/core/content/utils/annotation.py

```diff
@@ -30,109 +34,368 @@
+            tree = node[0]
+            lft = min_val = node[2]
+            rght = max_val = node[3]
+
+            if next_lft is None:
+                next_lft = rght + 1
+                min_max = {"min": min_val, "max": max_val}
+            elif lft == next_lft:
+                if min_val < min_max["min"]:
+                    min_max["min"] = min_val
+                if max_val > min_max["max"]:
+                    min_max["max"] = max_val
+                next_lft = rght + 1
+            elif lft != next_lft:
+                or_queries.append(
+                    and_(
+                        ContentNodeTable.c.tree_id == tree,
+                        ContentNodeTable.c.lft >= min_max["min"],
+                        ContentNodeTable.c.rght <= min_max["max"],
+                    )
+                )
+                min_max = {"min": min_val, "max": max_val}
+                next_lft = rght + 1
+        or_queries.append(
+            and_(
+                ContentNodeTable.c.tree_id == tree,
+                ContentNodeTable.c.lft >= min_max["min"],
+                ContentNodeTable.c.rght <= min_max["max"],
+            )
+        )
+
+    if not or_queries:
+        # No constraints that apply in this range, so therefore this query should always
+        # evaluate to False, because nothing can match it.
+        return select([ContentNodeTable.c.id]).where(False == True)  # noqa E712
+
+    # Return a query that ors each of the constraints
+    return select([ContentNodeTable.c.id]).where(or_(*or_queries))
+
 
 def set_leaf_nodes_invisible(channel_id, node_ids=None, exclude_node_ids=None):
+    """
+    Set nodes in a channel as unavailable.
+    With no additional arguments, this will hide an entire channel.
+    With the additional nodes arguments, it will selectively flag nodes
+    as unavailable, based on the passed in ids, setting them as unavailable if
+    they are in node_ids, or descendants of those nodes, but not in
+    exclude_node_ids or descendants of those nodes.
+    """
     bridge = Bridge(app_name=CONTENT_APP_NAME)
 
+    # SQL Alchemy reference to the content node table
     ContentNodeTable = bridge.get_table(ContentNode)
 
     connection = bridge.get_connection()
 
+    # Initial statement for nodes to update
     update_statement = ContentNodeTable.update().where(
         and_(
             ContentNodeTable.c.kind != content_kinds.TOPIC,
             ContentNodeTable.c.channel_id == channel_id,
         )
     )
+    # To chunk the tree, we first find the full extent of the tree - this gives the
+    # highest rght value for this channel.
+    max_rght = connection.execute(
+        select([func.max(ContentNodeTable.c.rght)]).where(
+            ContentNodeTable.c.channel_id == channel_id,
+        )
+    ).scalar()
+
+    # Start a counter for the while loop
+    i = 0
+
+    # Count the total number of constraints
+    constraint_count = len(node_ids or []) + len(exclude_node_ids or [])
```

Comment:
It's `None` because: https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html - but I think you have perhaps just temporarily forgotten this.

Reply:
He knew it originally, but then the memory was modified as a side effect of another call to the same function! /s

Reply:
> The first time that the function is called [without an argument], Python creates a persistent list. 

Wow, I had no idea that was the case!

I assume then that another thing I've done in the past has not been doing what I thought:

```python
def func(param=calc_default_value()):
```